iib/workers/tasks/build.py

```diff
@@ -337,6 +342,166 @@ def _get_resolved_image(pull_spec):
+        error occured.
+    :raises AddressAlreadyInUse: if the specified port is already being used by another service.
+    """
+    cmd = ['opm', 'registry', 'serve', '-p', str(port), '-d', db_path]
+    for attempt in range(max_tries):
+        rpc_proc = subprocess.Popen(
+            cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE, universal_newlines=True
+        )
+        time.sleep(wait_time)
+        ret = rpc_proc.poll()
+
+        # process has terminated
+        if ret is not None:
+            stderr = rpc_proc.stderr.read()
+            if 'address already in use' in stderr:
+                raise AddressAlreadyInUse(f'Port {port} is already used by a different service')
+            raise IIBError(f'Command "{" ".join(cmd)}" has failed with error "{stderr}"')
+
+        stdout = get_running_subprocess_output(rpc_proc.stdout)
+        if 'serving registry' in stdout:
+            log.debug('Started the command "%s"', ' '.join(cmd))
+            log.info('Index registry service has been initialized.')
+            return rpc_proc
+
+        rpc_proc.kill()
+
+    raise IIBError(f'Index registry has not been initialized after {max_tries} tries')
+
+
+@retry(attempts=2, wait_on=IIBError, logger=log)
+def _get_present_bundles(from_index, base_dir):
+    """
+    Get a list of bundles already present in the index image.
+
+    :param str from_index: index image to inspect.
+    :param str base_dir: base directory to create temporary files in.
+    :return: list of present bundles as provided by the grpc query.
+    :rtype: list
+    :raises IIBError: if any of the commands fail.
+    """
+    db_path = _get_index_database(from_index, base_dir)
+    port, rpc_proc = _serve_index_registry(db_path)
+
+    bundles = run_cmd(
+        ['grpcurl', '-plaintext', f'localhost:{port}', 'api.Registry/ListBundles'],
+        exc_msg='Failed to get bundle data from index image',
+    )
+    rpc_proc.kill()
+
+    # Transform returned data to parsable json
+    present_bundles = [json.loads(bundle) for bundle in re.split(r'(?<=})\n(?={)', bundles)]
+    return present_bundles
+
+
+@retry(attempts=2, wait_on=IIBError, logger=log)
+def _get_missing_bundles(present_bundles, bundles):
+    """
+    Filter out bundles to only those not present in the index image.
+
+    :param list present_bundles: list of bundles present in the index image, as provided by opm.
+    :param list bundles: bundles requested to be added to the index image.
+    :return: list of bundles not present in the index image.
+    :rtype: list
+    """
+    # we only need package name and version
+    present_bundle_ids = []
+    present_bundle_ids_path = []
+    present_bundle_hashes = []
+    for bundle in present_bundles:
+        present_bundle_ids.append((bundle['packageName'], bundle['version']))
+        # bundlePath could be defined either by hash or tag. Regardless, we'll use it for matching
```

Comment:
Could the tag be a floating tag? If so, we wouldn't want to match against that. I don't have enough context to know for sure.

I imagine that this is assuming that the tag is an NVR, however, that is an implementation specific to the internal Red Hat pipeline and not generic.

Reply:
Should I remove this? Should we ask someone else?